lib/spack/spack/environment.py

```diff
@@ -1509,6 +1509,45 @@ def concretized_specs(self):
         for s, h in zip(self.concretized_user_specs, self.concretized_order):
             yield (s, self.specs_by_hash[h])
 
+    def matching_spec(self, spec):
+        """
+        Given a spec (likely not concretized), find a matching concretized
+        spec in the environment. The matching spec does not have to be
+        installed in the environment.
+        """
+        root_matches = list()  # These are abstract
+        dep_matches = list()  # These are concrete
+        # These are concrete (for root specs, they store the concretized spec
+        # rather than the abstract spec)
+        all_matches = list()
+        for user_spec, concretized_user_spec in self.concretized_specs():
+            if concretized_user_spec.satisfies(spec):
+                root_matches.append(user_spec)
+                all_matches.append(concretized_user_spec)
+            for dep_spec in concretized_user_spec.traverse(root=False):
+                if dep_spec.satisfies(spec):
+                    dep_matches.append(dep_spec)
+                    all_matches.append(dep_spec)
+
+        if len(all_matches) == 1:
+            return all_matches[0]
+        elif len(all_matches) > 1:
+            # If multiple root specs match, it is assumed that the abstract
+            # spec will most-succinctly summarize the difference between them
+            # (and the user can enter one of these to disambiguate)
+            root_matches_str = '\n'.join(
+                "Root spec: {0}".format(str(x)) for x in root_matches)
+            # For dependencies of concretized specs, no summarization is
+            # available
+            dep_matches_str = '\n'.join(
+                "Dependency spec: {0}".format(str(x)) for x in dep_matches)
+            matches_str = '\n'.join(
+                x for x in [root_matches_str, dep_matches_str] if x)
+            msg = ("{0} matches multiple specs in the environment {1}: \n"
+                   "{2}".format(str(spec), self.name, matches_str))
+            raise SpackEnvironmentError(msg)
+        # If there are no matches, return nothing
```

Comment:
I think it's better to explicitly return `None` if we're relying on that behavior.